### DISCUSSION TITLE
Raspberry Pi binaries?

### DISCUSSION THREAD

**Discussion by mro** [helix-editor/helix#5841] [Q&A] [answered]
namely armhf linux. 

Are there plans to release such?

**Comment by the-mikedavis** [comment, marked as answer]
We provide arm64 linux builds in the [releases page](https://github.com/helix-editor/helix/releases/latest) (`helix-<version>-aarch64-linux.tar.xz`). From what I understand, some of the recent Raspberry Pi boards run arm64 (ref: https://www.raspberrypi.com/news/raspberry-pi-os-64-bit/).

armhf seems a bit niche for us to provide pre-built releases IMO. https://github.com/helix-editor/helix/issues/2171 talks about adding a snap for Helix though which might cover armhf if the snapcraft build service still can provide armhf builders.

**Comment by mro** [comment]
lol, armhf if niche, indeed, – so is terminal text editors, isn't it? :-).

But seriously, armhf is still the default for sold-today raspberry pi hardware. aarch64 is an option, but far from common or recommended. Raspberry has a strong backward compatibility tradition so you can flip SD-cards from one hardware to another from 1st gen to most recent.

When choosing an editor, muscle memory is a thing and so I'd prefer to have the editor on all the machines. And quite some are raspberry pis as I work also in an educational context where those are the kids workstations.

Snap isn't an option, I think. Binaries are.

Maybe someone from the community wants to contribute them.

**Reply by pascalkuthe** [reply]
If using helix on niche hardware is important for you, you can simply compile helix yourself. Either directly on your rasperry pi or using cross compilation

**Reply by mro** [reply]
I dread the rust toolchain.

**Reply by pascalkuthe** [reply]
why? rust has one of the easiest to work with toolchains of any system programming language with great cross compilation support. For building directly on the board you can just install rustup. The difficult part about cross compilation is getting an appropriate C compiler/linker and glibc. Many distros ship packages for that tough and if you develop for such a board you are likely already have if installed. If not [cargo cross](https://github.com/cross-rs/cross) is pretty great. In any case doing it yourself is the best way to get this done. Unless some dependency needs patching I doubt that it would be very difficult.

**Reply by CptPotato** [reply]
> Unless some dependency needs patching I doubt that it would be very difficult.

It's been a few months since then, but compiling on an RPi 2 was pretty much hassle free - other than the 70m compile time 🙃

**Reply by pascalkuthe** [reply]
> > Unless some dependency needs patching I doubt that it would be very difficult.
> 
> It's been a few months since then, but compiling on an RPi 2 was pretty much hassle free - other than the 70m compile time 🙃

That's why I sugessted cross compiling. On my pc helix compiles in 2-3min (from scrstch in release mode). Mutlicore amd64 CPUs really flex their muscle when compiling :D
 Good to know that all dependencies work correctly.gor that target too😃 32bit is pretty niche these days

**Reply by archseer** [reply]
> Maybe someone from the community wants to contribute them.

Maybe that could be... you?

Sometimes you need to be able to manually compile packages from your device.. especially if it's a raspberry pi. I've done so for my pi devices before.

**Reply by pickfire** [reply]
> That's why I sugessted cross compiling. On my pc helix compiles in 2-3min (from scrstch in release mode). Mutlicore amd64 CPUs really flex their muscle when compiling :D

I was using a pi 2 for ~4 years as main desktop back then probably 8-10 years back then, I don't even have another computer to cross-compile, I think cross-compile is not a luxury for everyone. T_T

**Comment by mro** [comment]
e.g. the alpine folks seem to do what I am out for https://pkgs.alpinelinux.org/packages?name=helix&arch=armhf

P.S.: I was at no time interested in lectures on toolchains. No matter how convenient or interesting they may appear to you. They are not for me. I am interested in an editor.

**Reply by dead10ck** [reply]
Wow dude. Pascal was trying to help you. Nobody is obligated to make a helix package for you. Check your entitlement.

**Reply by mro** [reply]
I am sorry that you took my question for an obligation. I expected nothing but also wanted no new hobby. I am very grateful that you build such an awesome editor and am happy if there are binaries to use.

**Comment by mro** [comment]
http://dl-cdn.alpinelinux.org/alpine/latest-stable/community/armhf/helix-22.08.1-r0.apk is a .tar.gz with usr/bin/hx.

**Comment by mro** [comment]
@pascalkuthe just to give you an idea why I said 'i dread the rust toolchain' earlier on.

After running `$ curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh` and adding `~/.cargo/bin` to `$PATH`, I get

```sh
$ rustc --version
error: command failed: 'rustc': No such file or directory (os error 2)
```

things look fishy, if you ask me:

```
$ rustup show
Default host: aarch64-unknown-linux-gnu
rustup home:  /home/mro/.rustup

stable-aarch64-unknown-linux-gnu (default)
(error reading rustc version)
```

**Reply by mro** [reply]
next try was the OS-provided `cargo`:

```sh
$ sudo apt-get install cargo
$ date ; cargo install --locked --path helix-term ; date
Di 21. Mär 14:50:53 CET 2023
error: failed to parse manifest at `/home/mro/Downloads/helix/helix-term/Cargo.toml`

Caused by:
  failed to parse the `edition` key

Caused by:
  this version of Cargo is older than the `2021` edition, and only supports `2015` and `2018` editions.
Di 21. Mär 14:50:53 CET 2023
```

**Reply by mro** [reply]
oh, and installing rust via OS and cargo manually gives:

```sh
$ curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
info: downloading installer
warning: it looks like you have an existing installation of Rust at:
warning: /bin
warning: rustup should not be installed alongside Rust. Please uninstall your existing Rust first.
warning: Otherwise you may have confusion unless you are careful with your PATH
warning: If you are sure that you want both rustup and your already installed Rust
warning: then please reply `y' or `yes' or set RUSTUP_INIT_SKIP_PATH_CHECK to yes
warning: or pass `-y' to ignore all ignorable checks.
error: cannot install while Rust is installed

Continue? (y/N)
```

**Reply by k12ish** [reply]
I suffered through this exact problem last night. When running `curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh` you get an interactive installer which gives you the option to tweak the reccommended installation.

Change the compilation host to `arm-unknown-linux-gnueabihf`!

https://stackoverflow.com/questions/67148415/fail-to-run-cargo-after-install-rust-on-raspberry-pi-4

**Comment by blthayer** [comment]
In case it's helpful to anyone else, here's my recent experience (`helix` repo at a62d374fcb489f83bddd3a775df250ff4b9b34a9):

On a Raspberry Pi Model 3B (_not_ 3B+) running the latest OS (Raspberry Pi OS (32-bit), 21 Apr 2026 release based on Debian 13 (trixie)), building helix from source on the Pi was maxing out both memory and swap, so I dove into cross-compiling. This is what worked for me on my x86_64 machine running Pop!_OS 22.04 LTS (should work on modern-ish Debian/Ubuntu as well):

Tooling (assuming you already have `rustup`):
```console
rustup target add armv7-unknown-linux-gnueabihf
sudo apt update
sudo apt install gcc-arm-linux-gnueabihf
```

Now, in the `helix` repository, add the following to `.cargo/config.toml`:

```toml
[target.armv7-unknown-linux-gnueabihf]
linker = "arm-linux-gnueabihf-gcc"
```

We're now ready to build. From the `helix` repo:

```console
HELIX_DISABLE_AUTO_GRAMMAR_BUILD=1 cargo build --profile opt --target armv7-unknown-linux-gnueabihf --package helix-term --locked
```

(I skipped the grammars because I already had all but a couple failed grammars on my Pi due to initial source build attempts)

You can find the binary at `target/armv7-unknown-linux-gnueabihf/opt/hx`. Seems to work on my Pi!

PS - FWIW, the source build worked fine on my Raspberry Pi 4 Model B Rev 1.5 running the 64 bit Trixie-based Debian.
